framework/include/materials/DerivativeMaterialInterface.h

```diff
@@ -63,7 +64,7 @@ class DerivativeMaterialInterface : public T
    * Fetch a pointer to a material property if it exists, otherwise return null
    */
   template<typename U>
-  MaterialProperty<U> * getMaterialPropertyPointer(const std::string & name);
+  MaterialProperty<U> * getMaterialPropertyPointer(const std::string & name) __attribute__ ((deprecated));
```

Comment:
> Btw. this syntax is supported by GCC, clang, and Intel.

Yes, and from what I've read it probably has been for quite some time.  Still, an important difference between `__attribute__` and compiler `#pragma`s is [apparently](http://stackoverflow.com/a/1458684) that the latter is simply ignored if not understood while the former causes a compile-time failure... so your macro idea (can you put a macro in between a function parameter list and function body?) would be the right approach if we decide to use it more extensively.